app/models/composition_player.rb

```diff
@@ -0,0 +1,31 @@
+class CompositionPlayer < ApplicationRecord
+  belongs_to :composition
+  belongs_to :player
+
+  before_validation :set_position
+
+  validates :player, :composition, :position, presence: true
+  validates :position, numericality: { only_integer: true }
+  validate :composition_does_not_have_max_players
+
+  default_scope { order(:position) }
+
+  private
+
+  def set_position
+    return unless composition
+
+    self.position ||= composition.composition_players.count
```

Comment:
The goal wasn't memoization here but rather to set a default value only if the caller didn't explicitly pass a position. I want the caller to be able to override the position, and that's what this allows: if `position` is already non-nil, it doesn't get set to the default value of how many other players are in this composition.

> the CompositionPlayer records created for me always have a position of 0

Were you using the same composition for each new CompositionPlayer you created? If not, I'd expect it to always be 0.

There's a model test, `sets position`, that shows the position defaults to something >0 as more players are added the composition.